### ISSUE TITLE
Crash On Incorrect Data Type Passed To Function

### ISSUE THREAD

**Issue by santorac** [microsoft/DirectXShaderCompiler#4691]
### Title
Crash On Incorrect Data Type Passed To Function

### Functional impact
It is difficult to track down the incorrect shader code because error messages are not reported.

### Minimal repro steps

[Skin_ForwardPass_dx12.hlsl.prepend.txt](https://github.com/microsoft/DirectXShaderCompiler/files/9658397/Skin_ForwardPass_dx12.hlsl.prepend.txt)

1. Use the attached  file "Skin_ForwardPass_dx12.hlsl.prepend.txt", rename to "Skin_ForwardPass_dx12.hlsl.prepend".
2. Run a command line like this (with adjusted paths as needed): "D:\o3de\build\bin\profile\Builders\DirectXShaderCompiler\dxc.exe" -E PixelShader -T ps_6_2 -Zpr -enable-16bit-types -Fo "D:\o3de\AtomSampleViewer\user\AssetProcessorTemp\JobTemp-QuNqiC\Skin_ForwardPass_dx12.dxil.bin" -Fh "D:\o3de\AtomSampleViewer\user\AssetProcessorTemp\JobTemp-QuNqiC\Skin_ForwardPass_dx12.dxil.txt" "D:/o3de/AtomSampleViewer/user/AssetProcessorTemp/JobTemp-QuNqiC/Skin_ForwardPass_dx12.hlsl.prepend"
3. Output: Internal compiler error: access violation. Attempted to read from address 0x0000000000000018
4. Now edit the file to fix the data type...
    1. Comment out line 10615, function call to "ApplyIBL"
    2. Uncomment line 10614: ReflectionProbeData temp; ApplyIBL ( surface , lightingData, temp, ::ObjectSrg_m_reflectionProbeCubeMap ) ;
5. Run the same dxc command again
6. Compilation is successful.

### Expected result
There should be an error message about the fact that an object of the wrong data type is being passed to ApplyIBL().

### Actual result
Crash
Internal compiler error: access violation. Attempted to read from address 0x0000000000000018

### Further technical details
Just FYI, this file was generated as an intermediate source file from a feature branch of https://github.com/o3de/o3de 

DXC Version: dxcompiler.dll: 1.6 - 1.6.0.3371 (release-1.6.2112-o3de, 5d3acf7ba); dxil.dll: 1.5(10.0.19041.685)

**Comment by pow2clk** [comment]
Hi @santorac!

As a matter of fact, this should succeed on HLSL versions prior to 2021. The rules before that version allowed struct/class parameters to be considered castable if their data members were all the same. However, we have a bug here that relates to the fact that they are declared as classes instead of structs. If you change them to structs, this should compile fine. 

Another alternative would be to switch to HLSL version 2021 using `-HV 2021`. This will give you the error you expected to see. However, you will then run into issues related to the short-circuiting changes in this particular shader. They should be easy to fix though. See https://devblogs.microsoft.com/directx/announcing-hlsl-2021/#logical-operator-short-circuiting for details. 

For reference and clarity, here is the minimal repro shader I came up with:

```hlsl
class A {
  float f;
  int i;
};
class B {
  float f;
  int i;
};

void badCall(B data) {}

float4  main() : SV_Target {
  A dnc;
  badCall(dnc);
  return 0.0;
}
```

**Comment by santorac** [comment]
I see, thanks for that info. I've worked around this issue with some updates to my shader code already, but just wanted to raise awareness about the crash.

I'm sure others will appreciate your reduced repro case instead of my 10K line one haha!
